src/System.Device.Gpio/System/Device/Gpio/Drivers/LibgpiodCapiDriver.cs

```diff
@@ -0,0 +1,269 @@
+
+        public string GpioChip { get; set; } = "gpiochip0";
+
+        private SafeChipHandle _chip;
+
+        private Dictionary<int, SafeLineHandle> _pinNumberToSafeLineHandle;
+
+        private timespec _defaultTimeOut = new timespec(IntPtr.Zero, new IntPtr(1_000_000)); // one millisecond
+
+        private Dictionary<int, LibgpiodDriverEventHandler> _pinNumberToEventHandler = new Dictionary<int, LibgpiodDriverEventHandler>();
+
+        protected internal override int PinCount
+        {
+            get
+            {
+                return Interop.gpiod_chip_num_lines(_chip);
+            }
+        }
+
+        public LibgpiodCapiDriver()
+        {
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        public LibgpiodCapiDriver(string chip)
+        {
+            if (chip != null)
+                GpioChip = chip;
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        private void OpenChip()
+        {
+            _chip = Interop.gpiod_chip_open($"{BasePath}{GpioChip}");
+            if (_chip == null)
+            {
+                throw new GpioChipException($"Chip: {GpioChip} not available, error code: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        private void RequestEvent(int pinNumber, LibgpiodDriverEventHandler eventDescriptor, PinEventTypes eventType)
+        {
+            int eventSuccess = -1;
+            SafeLineHandle pin = Interop.gpiod_chip_get_line(_chip, pinNumber);
+            if (pin != null)
+            {
+                eventDescriptor.PinHandle = pin;
+                if (eventType == PinEventTypes.Rising || eventType == PinEventTypes.Falling)
+                    eventSuccess = Interop.gpiod_line_request_both_edges_events(pin, $"Listen {pinNumber} for falling edge event");
+                else
+                    throw new GpioChipException("Invalid or not supported event type requested");
+            }
+            if (eventSuccess < 0)
+            {
+                throw new GpioChipException($"Error while requesting event listener for pin {pinNumber}, error code: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
+        {
+            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibgpiodDriverEventHandler eventHandler))
+            {
+                eventHandler = new LibgpiodDriverEventHandler(pinNumber, new CancellationTokenSource());
+                _pinNumberToEventHandler.Add(pinNumber, eventHandler);
+                RequestEvent(pinNumber, eventHandler, eventType);
+                InitializeEventDetectionThread(eventHandler);
+            }
+            if (eventType.HasFlag(PinEventTypes.Rising))
```

Comment:
For what it's worth, in high performance code you do not use `enum.HasFlag` because it is slower than a bitwise check (ie `eventType & PinEventTypes.Rising != 0`)

If it's not performance sensitive code (and I'm guessing this isn't) then it's a matter of style. CoreFX almost exclusively avoids HasFlag.

Reply:
I may be in a minority but I find HasFlag less readable...

Reply:
Good catch.

Reply:
This condition style copied from UnixDriver class, I guess I don't need to update that one